rfcs/text/0000-data_stream-fields.md

```diff
@@ -0,0 +1,181 @@
+-->
+
+## Usage
+
+<!--
+Stage 1: Describe at a high-level how these field changes will be used in practice. Real world examples are encouraged. The goal here is to understand how people would leverage these fields to gain insights or solve problems. ~1-3 paragraphs.
+-->
+
+Data stream fields are already in use in Elastic Agent. Leveraging the data stream fields described here allow users to filter by a specific data type (logs, metrics etc.), dataset (nginx.access, prometheus) or namespace. The following are examples of common queries pertaining to specific datatypes, datasets or namespaces:
+
+* `data_stream.type: logs`
+* `data_stream.dataset: nginx.access`
+* `data_stream.type: logs AND data_stream.namespace: web-frontend`
+
+As previously described, fields mapped as `constant_keyword` allows Elasticsearch to drastically optimize queries involving those fields. See the [Elasticsearch documentation](https://www.elastic.co/guide/en/elasticsearch/reference/current/faster-filtering-with-constant-keyword.html) on `constant_keyword` for more information.
+
+
+## Source data
+
+<!--
+Stage 1: Provide a high-level description of example sources of data. This does not yet need to be a concrete example of a source document, but instead can simply describe a potential source (e.g. nginx access log). This will ultimately be fleshed out to include literal source examples in a future stage. The goal here is to identify practical sources for these fields in the real world. ~1-3 sentences or unordered list.
+-->
+
+Today, Elastic Agent adds the data_stream fields in all documents ingested. It's also possible to use the fields in data from other data sources. Elasticsearch 7.9+ ships with built-in index template mappings which will ensure that documents indexed into data streams that match `logs-*-*` and `metrics-*-*` will get the fields mapped correctly to `constant_keyword` types. 
+
+### Using data_stream fields with regular indices
+`data_stream` fields only make sense when indexing into data streams. They should not to be used for regular indices.
+
+
+<!--
+Stage 2: Included a real world example source document. Ideally this example comes from the source(s) identified in stage 1. If not, it should replace them. The goal here is to validate the utility of these field changes in the context of a real world example. Format with the source name as a ### header and the example document in a GitHub code block with json formatting.
+-->
+
+<!--
+Stage 3: Add more real world example source documents so we have at least 2 total, but ideally 3. Format as described in stage 2.
+-->
+
+## Scope of impact
+
+* We've described that `generic` is a valid value for `data_stream.dataset` in some cases. Since `event.dataset` should always have the same value, this will also apply to `event.dataset`. We should update the documentation on `event.dataset` to reflect this.
```

Comment:
Additional point on this, and I think we'll be good here.

Up until now, `event.dataset` has not had any restrictions on it. Users with custom data sources have been using it to identify their datasets as well, but may not have been following Elastic's conventions.

This means users may need to translate their custom dataset values (e.g. `event.dataset: firewall/config`) to an equivalent legal dataset, according with the character restrictions imposed by the use of the value in an index name (`data_stream.dataset: firewall.config`).

I'm not sure if we'll need to be able to depend on a hard `data_stream.dataset = event.dataset` or if a sanitized equivalence is acceptable.

I think for now we should mention that the char restriction on `data_stream.dataset` means that users *may* have to translate their `event.dataset` to be a legal `data_stream.dataset`.

Reply:
good point, will add it.